### PR TITLE
Fix OutOfBoundsException when importing multiple entries in medline format

### PR DESCRIPTION
**Pull request by tschechlovdev**
@zellerdev found out a bug when importing more than one entry in the medline format into JabRef. 
JabRef version
JabRef 3.5
windows 8.1 6.3 amd64 
Java 1.8.0_25

Steps to reproduce:
1. Select more than entry from [here] (http://www.ncbi.nlm.nih.gov/pubmed) .
2. Select "Send to File" and then choose "Medline" Format.
3. Open Jabref and select "Import into new database".
4.  I get an IndexOutOfBoundException shown on the console and no entries are imported.

Alternatively use the testfiles I added in this PR.

Quick fix: Add an if clause around the substring method, that causes the Exception.

- [x] Tests created for changes

### CODE REVIEW

**Comment by boceckts** [comment]
I can confirm the bug but next time please create an issue for this and assign it to yourself if you want to fix it. Also please use the "copy version to clipboard" functionallity in the About dialog in the help menu to easily paste your JabRef-, OS- and Java-version.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 1]
```diff
@@ -127,91 +126,93 @@ public ParserResult importDatabase(BufferedReader reader) throws IOException {
                     }
                 }
                 String entry = current.toString();
+                if (entry.contains("-")) {
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: I think the while-loop above this is already trying to check the current line for validity and tries to skip all the invalid ones. I would propose to extract a new method called `private boolean checkLineValidity(String line)` and do all the checks in there such as: empty line, line doesn't contain "-", "-" symbol not at 4th position, etc...
then the while loop above can also be replaced and you don't need to modify the counter variable "j". You can then simply skip any invalid lines with:
```
if (!checkLineValidity(lines[j])) {
    continue;
}
```

Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: This would also mean you don't need to modify as much code as you did now because then you don't have the if-clause around all the code.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 2]
```diff
@@ -26,7 +26,6 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.regex.Pattern;
-
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: Add the empty line again to fix Travis build.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java (hunk 1 above)
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: But I wouldn't extract the while loop above, because in the while loop are also correcctions, like adding space if '-' is at third position. And putting the while loop into a method that returns a boolean could have side effects, since you don't expect, that this method is changing something.

Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: You could also extract to multiple methods, `private boolean first checkLineValidity(String line)`, if false try to correct it by calling another method that tries to insert spaces or similar things.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 3]
```diff
@@ -227,13 +231,20 @@ public ParserResult importDatabase(BufferedReader reader) throws IOException {
...
         }
 
         return new ParserResult(bibitems);
 
     }
 
+    private boolean checkLineValidity(String line) {
+        if (line.length() >= 4) {
+            return line.contains("-") || !(line.charAt(4) == '-') ? true : false;
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: `!(line.charAt(4) == '-')` this will throw an OutOfBoundsException if the length of the line is exactly 4. Please cover this case.

Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: Also please check the if statement again, I can't find a case where this will return `false`.

Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: `? true : false` is unnecessary.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 4]
```diff
@@ -212,6 +215,7 @@ public ParserResult importDatabase(BufferedReader reader) throws IOException {
                     }
                     comment = comment + value;
                 }
+
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: Why did you add this line?

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 5]
```diff
@@ -227,13 +231,20 @@ public ParserResult importDatabase(BufferedReader reader) throws IOException {
             // create one here
             b.setField(fields);
             bibitems.add(b);
-
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: Why did you remove this line?

Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: because it was an unnecessary line.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 6]
```diff
@@ -238,11 +237,7 @@ public ParserResult importDatabase(BufferedReader reader) throws IOException {
     }
 
     private boolean checkLineValidity(String line) {
-        if (line.length() >= 4) {
-            return line.contains("-") || (line.charAt(4) != '-') ? true : false;
-        } else {
-            return false;
-        }
+        return (line.length() >= 5) ? line.contains("-") && (line.charAt(4) == '-') : false;
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: I think `line.contains("-")` is unnecessary when you also check for it's position.

File: src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java [hunk 7]
```diff
@@ -237,7 +237,7 @@ public ParserResult importDatabase(BufferedReader reader) throws IOException {
     }
 
     private boolean checkLineValidity(String line) {
-        return (line.length() >= 5) ? line.contains("-") && (line.charAt(4) == '-') : false;
+        return (line.length() >= 5) ? (line.charAt(4) == '-') : false;
```
Comment (matthiasgeiger) [inline src/main/java/net/sf/jabref/importer/fileformat/MedlinePlainImporter.java]: `return ((line.length() >= 5) && (line.charAt(4) == '-'))`  - why not as simple as that? :wink:

**Comment by tschechlovdev** [comment]
No remarks?

File: src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java [hunk 8]
```diff
@@ -198,6 +198,14 @@ public void testWithNbibFile() throws IOException, URISyntaxException {
     }
 
     @Test
+    public void testWithMultipleEntries() throws IOException, URISyntaxException {
+        Path file = Paths
+                .get(MedlinePlainImporter.class.getResource("MedlinePlainImporterStringOutOfBounds.txt").toURI());
+        List<BibEntry> entries = importer.importDatabase(file, Charset.defaultCharset()).getDatabase().getEntries();
```
Comment (Siedlerchr) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java]: When the bib file is in UTF8, then please set the Charset explicit to UTF8, too. `DefaultCharsets.UTF8` (or similar), otherwise test will fail on windows in gradle.

**Comment by Siedlerchr** [comment]
In general good. :+1:  But I would recommend adding some test with invalid formats, e.g. if an exceptions is thrown (expected) or whatever.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java (hunk 8 above)
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java]: done.

**Comment by tschechlovdev** [comment]
Hm an Excpetion should be not thrown, except if the reader would be null. If the format is invalid then the entries list will be empty, but no exception will be thrown. I'll added tests for both.

**Comment by tschechlovdev** [comment]
Any other remarks?

**Comment by stefan-kolb** [comment]
LGTM, just one remark: can you keep the test files as minimal as needed inside the tests? No need for 300 attributes when you can recreate the test issue with one or two.

**Comment by tschechlovdev** [comment]
Ok, I've minimized the testfile for the issue.

**Comment by stefan-kolb** [comment]
Maybe fix the two Codacy warnings (https://www.codacy.com/app/simonharrer/jabref/pullRequest?prid=289709) and add a key to the @misc entries. Otherwise im :+1: We need a second review from a @JabRef/developers tho.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java [hunk 9]
```diff
@@ -201,11 +204,27 @@ public void testWithNbibFile() throws IOException, URISyntaxException {
...
+        List<BibEntry> entries = importer.importDatabase(file, Charsets.UTF_8).getDatabase().getEntries();
         BibEntryAssert.assertEquals(MedlinePlainImporter.class, "MedlinePlainImporterStringOutOfBounds.bib", entries);
     }
 
     @Test
+    public void testInvalidFormat() throws URISyntaxException, IOException {
+        Path file = Paths
+                .get(MedlinePlainImporter.class.getResource("MedlinePlainImporterTestInvalidFormat.xml").toURI());
+        List<BibEntry> entries = importer.importDatabase(file, Charsets.UTF_8).getDatabase().getEntries();
+        assertEquals(Collections.EMPTY_LIST, entries);
```
Comment (Siedlerchr) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java]: Use the method Collections,emptyList() rather than the constant, because the constant does not return a generic list and is therefore not typesafe.
(In this case not so important, but could possinly produce an unchecked warning)

File: src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java [hunk 10]
```diff
@@ -201,11 +204,27 @@ public void testWithNbibFile() throws IOException, URISyntaxException {
...
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void testNullReader() throws IOException {
+        try (BufferedReader reader = null) {
+            ParserResult entries = importer.importDatabase(reader);
```
Comment (Siedlerchr) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java]: For the codacy issue just remove the assignment

**Comment by Siedlerchr** [comment]
Just the collections,emptyList and the codacy and it's fine :+1:  ( I already reviewd the main parts before)

File: src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java (hunk 9 above)
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java]: done.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java (hunk 10 above)
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlinePlainImporterTest.java]: done.
